# **Importing Libraries and Reading Data**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats


In [ ]:
# Show all columns
pd.set_option('display.max_columns', None)

# Load the datasets
df_customer_feedback = pd.read_csv('/content/drive/MyDrive/3rd year 2024-25/Term 2/Data Mining Principles/raw data/Customer_Feedback_Data.csv')
df_product_offering = pd.read_csv('/content/drive/MyDrive/3rd year 2024-25/Term 2/Data Mining Principles/raw data/Product_Offering_Data.csv')
df_transaction = pd.read_csv('/content/drive/MyDrive/3rd year 2024-25/Term 2/Data Mining Principles/raw data/Transaction_Data.csv')


# **Data Overview**

In [ ]:
# Dictionary of DataFrames
dfs = {
    'Customer Feedback Data': df_customer_feedback,
    'Product Offering Data': df_product_offering,
    'Transaction Data': df_transaction
}

### Summary Statistics

In [ ]:
for name, df in dfs.items():
    print(f"{name}:\n{df.describe()}\n")

Customer Feedback Data:
       Customer_ID  Satisfaction_Score  Likelihood_to_Recommend
count  5050.000000         4949.000000              5050.000000
mean    501.949703            5.684785                 5.571287
std     288.806331            3.606511                 2.872577
min       1.000000            1.000000                 1.000000
25%     253.000000            3.000000                 3.000000
50%     502.000000            6.000000                 6.000000
75%     751.750000            8.000000                 8.000000
max    1000.000000           60.000000                10.000000

Product Offering Data:
       Product_ID  Target_Age_Group
count   15.000000               0.0
mean     5.133333               NaN
std      3.159265               NaN
min      1.000000               NaN
25%      2.500000               NaN
50%      5.000000               NaN
75%      7.500000               NaN
max     10.000000               NaN

Transaction Data:
       Transaction_ID  Customer_I

### Number of rows and columns in the DataFrame


In [ ]:
for name, df in dfs.items():
    print(f"{name}: {df.shape}")

Customer Feedback Data: (5050, 4)
Product Offering Data: (15, 6)
Transaction Data: (5050, 5)


###Display Index, Columns, and Data

In [ ]:
for name, df in dfs.items():
    print(f"\n=== {name} ===")
    print("DataFrame Info:\n")
    df.info()




=== Customer Feedback Data ===
DataFrame Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5050 entries, 0 to 5049
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Customer_ID              5050 non-null   int64  
 1   Satisfaction_Score       4949 non-null   float64
 2   Feedback_Comments        5050 non-null   object 
 3   Likelihood_to_Recommend  5050 non-null   int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 157.9+ KB

=== Product Offering Data ===
DataFrame Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Product_ID           15 non-null     int64  
 1   Product_Name         15 non-null     object 
 2   Product_Type         15 non-null     object 
 3   Risk_Level           15 non-null     object 
 4   Ta

In [ ]:
for name, df in dfs.items():
    print(f"{name}:\n{df.head()}\n")

Customer Feedback Data:
   Customer_ID  Satisfaction_Score  Feedback_Comments  Likelihood_to_Recommend
0            1                10.0     Very satisfied                        9
1            2                 3.0     Very satisfied                        3
2            3                10.0     Very satisfied                        1
3            4                 7.0  Needs improvement                        4
4            5                 8.0     Unsatisfactory                        7

Product Offering Data:
   Product_ID                   Product_Name     Product_Type Risk_Level  \
0           1           Platinum Credit Card      Credit Card     Medium   
1           2           Gold Savings Account  Savings Account        Low   
2           3  High-Yield Investment Account       Investment       High   
3           4                  Mortgage Loan             Loan     Medium   
4           5                      Auto Loan             Loan     Medium   

   Target_Age_Group T

##**Observations after Cursory look at the data**


####**Customer Feedback Data**
This dataset seems to be about collecting feedback from customers about their experiences with a product or service.

|Column Name|Description|Data Type|Sample Data|
|:---|:---|:---|:---|
|Customer_ID|A unique identifier for each customer|int|1, 2, 3|
|Satisfaction_Score|A numerical score indicating how satisfied the customer is|float|10.0, 3.0, 7.0|
|Feedback_Comments|A categorical field with short comments about the service|object|Very satisfied, Needs improvement|
|Likelihood_to_Recommend|A numerical score showing how likely the customer is to recommend the product/service|int|9, 3, 1|

Based on the df.describe():
*   `Customer_ID`: Ranges from 1-1000, but total count is 5050 so there are repeat customers that gave multiple feedback.
*   `Satisfaction_Score`: Max is 60, but mean is 5.68 and 75th prinicple is 8.00, 60 seems like an outlier.
*   `Likelihood_to_Recommend`: Range is 1 to 10, seems to be a scale.

Recommendations:


*   Change `Satisfaction_Score` to int.
*   `Feedback_Comments`, check if it can be categorical, seems like the same comments repeat. There are only 5 comments throughout the column,Very satisfied, Good service, Needs improvement, Unsatisfactory and Excellent . Is there correlation with `Satisfaction_Score` and/or `Likelihood_to_Recommend`?

####**Product Offering Data**
This dataset seems to be a catalog or list of products/services offered by a company.

|Column Name|Description|Data Type|Sample Data|
|:---|:---|:---|:---|
|Product_ID|A unique identifier for each product|int|1, 2, 3|
|Product_Name|The name of the product|object|Platinum Credit Card, Gold Savings Account|
|Product_Type|A broader category of the product (e.g., Credit Card, Loan)|object|Credit Card, Savings Account|
|Risk_Level|The risk level associated with the product|object|Medium, Low, High|
|Target_Age_Group|This column is empty, so its purpose is unclear without more context|float|||
|Target_Income_Group|The income level the product is designed for|object|Medium, Low, High|

Based on the df.describe():
*   `Product_ID`: Ranges from 1-10, but total count is 15 so there are either duplicates or multiple products with the same Product_ID.
*   `Target_Age_Group`: Whole column is missing data. Drop this column?
*   `Risk_Level` and `Target_Income_Group`: Both are scales. They can be processed using ordinal encoding (is this needed? hm..).

####**Transaction Data**
This dataset seems to be recorded transactions made by customers.

|Column Name|Description|Data Type|Sample Data|
|:---|:---|:---|:---|
|Transaction_ID|A unique identifier for each transaction|int|1, 2, 3|
|Customer_ID|The identifier of the customer who made the transaction|int|393, 826, 916|
|Transaction_Date|The date and time of the transaction|object|2023-01-01 00:00:00, 2023-01-01 01:00:00|
|Transaction_Amount|The amount of money involved in the transaction|float|3472.0, 10.0, 72.0|
|Transaction_Type|The type of transaction (e.g., Purchase, Bill Payment)|object|Purchase, Bill Payment, Investment|

Recommendations:
*   `Transaction_Date`: Change data type to datetime.
*   `Customer_ID`: Can be cross referenced with `Customer_ID` from Customer Feedback Data.

# **Data Cleaning**

In [ ]:
# Copies of original data set
feedback = df_customer_feedback.copy()
products = df_product_offering.copy()
transactions = df_transaction.copy()

### Removing Duplicates

In [ ]:
# --- Check for and remove duplicates in 'feedback' DataFrame ---

# Count duplicates
duplicates_feedback = feedback[feedback.duplicated()]
num_duplicates_feedback = len(duplicates_feedback)
print("Number of duplicate rows in feedback:", num_duplicates_feedback)

# Remove duplicates, keeping the first occurrence
feedback = feedback.drop_duplicates(keep='first')

# --- Check for and remove duplicates in 'products' DataFrame ---

# Count duplicates
duplicates_products = products[products.duplicated()]
num_duplicates_products = len(duplicates_products)
print("Number of duplicate rows in products:", num_duplicates_products)

# Remove duplicates, keeping the first occurrence
products = products.drop_duplicates(keep='first')

# --- Check for and remove duplicates in 'transactions' DataFrame ---

# Count duplicates
duplicates_transactions = transactions[transactions.duplicated()]
num_duplicates_transactions = len(duplicates_transactions)
print("Number of duplicate rows in transactions:", num_duplicates_transactions)

# Remove duplicates, keeping the first occurrence
transactions = transactions.drop_duplicates(keep='first')

# --- Print confirmation (optional) ---
print("\nDuplicates removed. DataFrames are now unique (keeping the first occurrence).")

Number of duplicate rows in feedback: 81
Number of duplicate rows in products: 5
Number of duplicate rows in transactions: 50

Duplicates removed. DataFrames are now unique (keeping the first occurrence).


### Removing outliers

In [ ]:
# Count the occurrences of each unique value in 'Satisfaction_Score'
satisfaction_counts = feedback['Satisfaction_Score'].value_counts()

# Print the results
print(satisfaction_counts)

Satisfaction_Score
7.0     532
5.0     523
10.0    505
6.0     497
8.0     485
4.0     485
9.0     469
3.0     469
2.0     454
1.0     440
60.0      2
51.0      2
58.0      2
52.0      1
56.0      1
53.0      1
54.0      1
Name: count, dtype: int64


Since there are only 10 outliers, it is better to drop these data, than replace as it could skew the results if kept.

In [ ]:
   # Identify outliers (scores above 10)
outlier_mask = feedback['Satisfaction_Score'] > 10

# Remove outlier rows
feedback = feedback[~outlier_mask]  # ~ inverts the boolean mask

# Verify that outliers are removed (optional)
print(feedback['Satisfaction_Score'].value_counts())

Satisfaction_Score
7.0     532
5.0     523
10.0    505
6.0     497
8.0     485
4.0     485
3.0     469
9.0     469
2.0     454
1.0     440
Name: count, dtype: int64


### Customer Feedback data Filling Missing Values

In [ ]:
# Count missing values for each column
missing_values = feedback.isnull().sum()

# Print the results
print(missing_values)

Customer_ID                  0
Satisfaction_Score         100
Feedback_Comments            0
Likelihood_to_Recommend      0
dtype: int64


101 our of 5050 is missing. Only 2% of data is missing.

In [ ]:
# Calculate the median Satisfaction_Score
median_satisfaction = feedback['Satisfaction_Score'].median()

# Fill missing values using .loc
feedback.loc[feedback['Satisfaction_Score'].isnull(), 'Satisfaction_Score'] = median_satisfaction

# (Optional) Verify that there are no more missing values
print(feedback['Satisfaction_Score'].isnull().sum())

0


### Product data Filling Missing Values

In [ ]:
# Count missing values for each column
missing_values = products.isnull().sum()

# Print the results
print(missing_values)

Product_ID              0
Product_Name            0
Product_Type            0
Risk_Level              0
Target_Age_Group       10
Target_Income_Group     0
dtype: int64


In [ ]:
# Drop the 'Target_Age_Group' column
products = products.drop('Target_Age_Group', axis=1)

### Transaction data Filling Missing Values

In [ ]:
# Count missing values for each column
missing_values = transactions.isnull().sum()

# Print the results
print(missing_values)

Transaction_ID          0
Customer_ID             0
Transaction_Date        0
Transaction_Amount    100
Transaction_Type        0
dtype: int64


In [ ]:
# Calculate the mean Transaction_Amount for each Transaction_Type
mean_by_type = transactions.groupby('Transaction_Type')['Transaction_Amount'].mean()

# Fill missing values in 'Transaction_Amount' using the calculated means
transactions['Transaction_Amount'] = transactions.apply(
    lambda row: mean_by_type[row['Transaction_Type']] if pd.isnull(row['Transaction_Amount']) else row['Transaction_Amount'],
    axis=1
)

# Verify that there are no more missing values
print(transactions['Transaction_Amount'].isnull().sum())

0


### Changing Data Types

In [ ]:
for name, df in dfs.items():
    print(f"Data types for {name}:")
    print(df.dtypes)
    print()


Data types for Customer Feedback Data:
Customer_ID                  int64
Satisfaction_Score         float64
Feedback_Comments           object
Likelihood_to_Recommend      int64
dtype: object

Data types for Product Offering Data:
Product_ID               int64
Product_Name            object
Product_Type            object
Risk_Level              object
Target_Age_Group       float64
Target_Income_Group     object
dtype: object

Data types for Transaction Data:
Transaction_ID          int64
Customer_ID             int64
Transaction_Date       object
Transaction_Amount    float64
Transaction_Type       object
dtype: object



In [ ]:
# Change Data types for Customer Feedback Data:

# Convert 'Satisfaction_Score' to integer
feedback['Satisfaction_Score'] = feedback['Satisfaction_Score'].astype(int)

# Convert 'Feedback_Comments' to category
feedback['Feedback_Comments'] = feedback['Feedback_Comments'].astype('category')

# Check data types
feedback.dtypes

Customer_ID                   int64
Satisfaction_Score            int64
Feedback_Comments          category
Likelihood_to_Recommend       int64
dtype: object

In [ ]:
# Change Data types for Product Offering Data:

# Convert 'Product_Type' to category
products['Product_Type'] = products['Product_Type'].astype('category')

# Convert 'Risk_Level' to category
products['Risk_Level'] = products['Risk_Level'].astype('category')

# Convert 'Target_Income_Group' to category
products['Target_Income_Group'] = products['Target_Income_Group'].astype('category')

# Check data types
products.dtypes

Product_ID                int64
Product_Name             object
Product_Type           category
Risk_Level             category
Target_Income_Group    category
dtype: object

In [ ]:
# Change Data types for Transaction Dat

# Convert 'Transaction_Date' to datetime
transactions['Transaction_Date'] = pd.to_datetime(transactions['Transaction_Date'])

# Convert 'Transaction_Type' to category
transactions['Transaction_Type'] = transactions['Transaction_Type'].astype('category')

# Check data types
transactions.dtypes

Transaction_ID                 int64
Customer_ID                    int64
Transaction_Date      datetime64[ns]
Transaction_Amount           float64
Transaction_Type            category
dtype: object

# **Feature Engineering?**

In [ ]:
# prompt: list unique values in transaction type and prodyct type

print(transactions['Transaction_Type'].unique())
print(products['Product_Type'].unique())

['Purchase', 'Bill Payment', 'Investment', 'Loan Payment']
Categories (4, object): ['Bill Payment', 'Investment', 'Loan Payment', 'Purchase']
['Credit Card', 'Savings Account', 'Investment', 'Loan']
Categories (4, object): ['Credit Card', 'Investment', 'Loan', 'Savings Account']
